metrics/internal/stackdriver/secret_watcher.go

```diff
@@ -0,0 +1,279 @@
+	createStackdriverKubeclientFunc func() (kubernetes.Interface, error)
+)
+
+func init() {
+	createStackdriverKubeclientFunc = createKubeclient
+}
+
+// NewSecretWatcher constructs a secretWatcher that watches all Secrets.
+func NewSecretWatcher(observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal("", "", observers...)
+}
+
+// NewSecretWatcherSingleNamespace constructs a secretWatcher that watches all Secrets in a namespace.
+func NewSecretWatcherSingleNamespace(namespace string, observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal(namespace, "", observers...)
+}
+
+// NewSecretWatcherSingleSecret constructs a secretWatcher that watches all a single Secret.
+func NewSecretWatcherSingleSecret(namespace string, name string, observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal(namespace, name, observers...)
+}
+
+// newSecretWatcherInternal constructs a secretWatcher.
+func newSecretWatcherInternal(namespace string, name string, observers ...Observer) (SecretWatcher, error) {
+	impl := &secretWatcherImpl{
+		SecretNamespace: namespace,
+		SecretName:      name,
+		observers:       observers,
+	}
+
+	if clientErr := impl.setupKubeclient(); clientErr != nil {
+		return nil, clientErr
+	}
+
+	return impl, nil
+}
+
+// secretWatherImpl implmements the secretWatcher interface.
+type secretWatcherImpl struct {
+	// kubeclient is the in-cluster Kubernetes kubeclient, which is lazy-initialized on first use.
+	kubeclient kubernetes.Interface
+
+	// watchLock ensures that only one watch is being run at a time.
+	watchLock sync.Mutex
+	// informer watches the Stackdriver secret if useStackdriverSecretEnabled is true.
+	informer cache.SharedIndexInformer
+	// stopCh stops the secretWatcherInformer when required.
+	stopCh chan struct{}
+	// watchStarted is whether or not the informer is running for the secret being watched.
+	watchStarted bool
+
+	// SecretName is the name of the secret being watched.
+	SecretName string
+	// SecretNamespace is the namespace of the secret being watched.
+	SecretNamespace string
+	// observers is the list of observers to trigger when the Secret is changed.
+	observers []Observer
+}
+
+// StartWatch implements StartWatch of secretWatcher interface.
+func (s *secretWatcherImpl) StartWatch() error {
+	s.watchLock.Lock()
+	defer s.watchLock.Unlock()
+	if s.watchStarted {
+		return nil
+	}
+	s.watchStarted = true
+
+	s.setupInformer()
+	s.stopCh = make(chan struct{})
+	go s.informer.Run(s.stopCh)
+
+	return nil
+}
+
+// StopWatch implements StopWatch secretWatcher interface.
+func (s *secretWatcherImpl) StopWatch() {
+	s.watchLock.Lock()
+	defer s.watchLock.Unlock()
+	if !s.watchStarted {
+		return
+	}
+	s.watchStarted = false
+
+	close(s.stopCh)
+}
+
+// GetSecret implements GetSecret of secretWatcher interface.
+func (s *secretWatcherImpl) GetSecret(namespace string, name string) (*corev1.Secret, error) {
```

Comment:
Is GetSecret necessary? Should the only interface be on the observer?

Reply:
It is not strictly necessary for a SecretWatcher, but it is convenient to include here to avoid duping the kubeclient initialization because it can reuse the same kubeclient as the SecretWatcher. 

Could you clarify what you mean by the only interface on the observer?

Reply:
Should the only API to receive the secret be the observer? Meaning the API does not support a non-observable way to fetch the secret. It ensures that consumers must respect that the secret value can change and build their usage accordingly.

Reply:
Ah, I see. That's an interesting idea, but it may be a bit limiting in usage. (Also, consumers can always cache the last Secret they saw from an observer which is potentially less desirable than allowing getting the most up-to-date value). 

In the case of watching the Stackdriver Secret, I would like check the Secret's existence with a Get before calling StartWatch. That way I can ensure (stackdriver secret configured && stackdriver secret exists) before burning any extra CPU to watch the Secret in the background.

Reply:
I understand the usecase of checking the configmap if the secret type is desired before creating this observable. But if the user specifies the stackdriver secret as desired, then an observer will need to be created eventually. Does the observer emit the current value set before the observer was created, or only on updates?

Reply:
The observer does get "create" notifications for all pre-existing secrets when the watch starts. I see your point that we should start the SecretWatch as long as "use-secret" configmap field is set. 

I'll remove the GetSecret func from the interface. It might create some disconnect for the user between when they update config-observability & when they actually see the metrics exporter get updated, but I think we can make that acceptable with some extra logging.